Python/bytecodes.c

```diff
@@ -2479,6 +2479,11 @@ dummy_func(
             PyFunctionObject *func = (PyFunctionObject *)callable;
             DEOPT_IF(func->func_version != func_version, CALL);
             PyCodeObject *code = (PyCodeObject *)func->func_code;
+            assert(func->func_defaults);
+            assert(PyTuple_CheckExact(func->func_defaults));
+            int defcount = (int)PyTuple_GET_SIZE(func->func_defaults);
+            assert(defcount <= code->co_argcount);
+            int min_args = code->co_argcount - defcount;
             DEOPT_IF(argcount > code->co_argcount, CALL);
```

Comment:
> Why do we need these checks at all? I suspect that I added them, but I'm not sure why.

It [looks like](https://github.com/python/cpython/pull/29033/files#diff-97ea1759f82531a45fa685065ff86bc1077c3ab3d5f65220eea48b7730032bc8R38-R42) we just had space for it at the time (when cache sizes were fixed).

> If `func->func_version == func_version` then the only variable is `argcount` which is `oparg + is_meth`. So we only need to know which of the two possible values of `is_meth` is safe. Can precompute those at specialization, so we only need to check `is_meth` at runtime.

The obvious way to do this would be to split this into two specializations: one that expects `is_meth`, and one that doesn't.

With that said, I'm not sure it's worth breaking up. There would be a decent amount of code duplication, and this opcode is relatively rare (~0.1% of all instructions executed). So maybe we should just leave it.

Maybe if we *really* wanted to, we could use the low bit of the `func_version` cache entry. Or, as you mentioned, we could encode the "shape" of the function's signature and use that instead of the version.